File: web/src/lib/api/jail/jail.ts
import { APIResponseSchema, type APIResponse } from '$lib/types/common';
import {
	JailLogsSchema,
	JailSchema,
	JailStateSchema,
	JailStatSchema,
	SimpleJailSchema,
	type CreateData,
	type Jail,
	type JailLogs,
	type JailStat,
	type JailState,
	type SimpleJail
} from '$lib/types/jail/jail';
import { apiRequest } from '$lib/utils/http';
import { z } from 'zod/v4';

export async function newJail(data: CreateData): Promise<APIResponse> {
	return await apiRequest('/jail', APIResponseSchema, 'POST', {
		name: data.name,
		ctId: parseInt(data.id.toString(), 10),
		description: data.description,
		dataset: data.storage.dataset,
		base: data.storage.base,
		switchId: data.network.switch,
		dhcp: data.network.dhcp,
		slaac: data.network.slaac,
		inheritIPv4: data.network.inheritIPv4,
		inheritIPv6: data.network.inheritIPv6,
		ipv4: data.network.ipv4,
		ipv4Gw: data.network.ipv4Gateway,
		ipv6: data.network.ipv6,
		ipv6Gw: data.network.ipv6Gateway,
		mac: data.network.mac,
		resourceLimits: data.hardware.resourceLimits,
		cores: parseInt(data.hardware.cpuCores.toString(), 10),
		memory: parseInt(data.hardware.ram.toString(), 10),
		startAtBoot: data.hardware.startAtBoot,
		startOrder: data.hardware.bootOrder
	});
}

export async function getSimpleJails(): Promise<SimpleJail[]> {
	return await apiRequest('/jail/simple', z.array(SimpleJailSchema), 'GET');
}

export async function getJails(): Promise<Jail[]> {
	return await apiRequest('/jail', z.array(JailSchema), 'GET');
}

export async function deleteJail(ctId: number): Promise<APIResponse> {
	return await apiRequest(`/jail/${ctId}`, APIResponseSchema, 'DELETE');
}

export async function getJailStates(): Promise<JailState[]> {
	return await apiRequest('/jail/state', z.array(JailStateSchema), 'GET');
}

export async function jailAction(ctId: number, action: string): Promise<APIResponse> {
	return await apiRequest(`/jail/action/${ctId}/${action}`, APIResponseSchema, 'POST');
}

export async function updateDescription(id: number, description: string): Promise<APIResponse> {
	return await apiRequest('/jail/description', APIResponseSchema, 'PUT', {
		id,
		description
	});
}

export async function getJailLogs(id: number, start: boolean): Promise<JailLogs> {
	return await apiRequest(`/jail/${id}/logs?start=${start}`, JailLogsSchema, 'GET');
}

export async function getStats(ctId: number, limit: number): Promise<JailStat[]> {
	return await apiRequest(`/jail/stats/${ctId}/${limit}`, z.array(JailStatSchema), 'GET');
}

export async function inheritHostNetwork(
	ctId: number,
	ipv4: boolean,
	ipv6: boolean
): Promise<APIResponse> {
	return await apiRequest('/jail/network/inheritance', APIResponseSchema, 'POST', {
		ctId,
		ipv4,
		ipv6
	});
}

export async function disinheritHostNetwork(ctId: number): Promise<APIResponse> {
	return await apiRequest(`/jail/network/disinherit/${ctId}`, APIResponseSchema, 'DELETE');
}

export async function addNetwork(
	ctId: number,
	switchId: number,
	macId: number,
	ip4: number,
	ip4gw: number,
	ip6: number,
	ip6gw: number,
	dhcp: boolean,
	slaac: boolean
): Promise<APIResponse> {
	return await apiRequest('/jail/network', APIResponseSchema, 'POST', {
		ctId,
		switchId,
		macId,
		ip4,
		ip4gw,
		ip6,
		ip6gw,
		dhcp,
		slaac
	});
}

export async function deleteNetwork(ctId: number, networkId: number): Promise<APIResponse> {
	return await apiRequest(`/jail/network/${ctId}/${networkId}`, APIResponseSchema, 'DELETE');
}

export async function updateResourceLimits(ctId: number, enabled: boolean): Promise<APIResponse> {
	return await apiRequest(
		`/jail/resource-limits/${ctId}?enabled=${enabled}`,
		APIResponseSchema,
		'PUT'
	);
}

File: internal/interfaces/services/jail/jail.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package jailServiceInterfaces

type CreateJailRequest struct {
	Name        string `json:"name" binding:"required"`
	CTID        *int   `json:"ctId" binding:"required"`
	Description string `json:"description"`
	Dataset     string `json:"dataset"`
	Base        string `json:"base"`

	SwitchId *int `json:"switchId"`

	InheritIPv4 *bool `json:"inheritIPv4"`
	InheritIPv6 *bool `json:"inheritIPv6"`

	DHCP  *bool `json:"dhcp"`
	SLAAC *bool `json:"slaac"`

	IPv4   *int `json:"ipv4"`
	IPv4Gw *int `json:"ipv4Gw"`

	IPv6   *int `json:"ipv6"`
	IPv6Gw *int `json:"ipv6Gw"`

	MAC *int `json:"mac"`

	Cores  *int `json:"cores"`
	Memory *int `json:"memory"`

	StartAtBoot *bool `json:"startAtBoot" binding:"required"`
	StartOrder  int   `json:"startOrder"`
}

type SimpleList struct {
	ID    uint   `json:"id"`
	Name  string `json:"name"`
	CTID  int    `json:"ctId"`
	State string `json:"state"`
}

type State struct {
	CTID      int    `json:"ctId"`
	State     string `json:"state"`
	PCPU      int64  `json:"pcpu"`
	Memory    int64  `json:"memory"`
	WallClock int64  `json:"wallClock"`
}

type JailServiceInterface interface {
	StoreJailUsage() error
}
